src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+        )
+
+        # Attention and hidden_states will be [] or None if they aren't needed
+        decoder_outputs: Tuple = _filter_out_falsey_values(decoder_outputs)
+        assert isinstance(decoder_outputs[0], tuple) # last layer hidden state of main stream and predicting streams
+        encoder_outputs: Tuple = _filter_out_falsey_values(encoder_outputs)
+        return decoder_outputs + encoder_outputs
+
+
+def _reorder_buffer(attn_cache, new_order):
+    for k, input_buffer_k in attn_cache.items():
+        if input_buffer_k is not None:
+            attn_cache[k] = input_buffer_k.index_select(0, new_order)
+    return attn_cache
+
+def _make_linear_from_emb(emb):
+    vocab_size, emb_size = emb.weight.shape
+    lin_layer = nn.Linear(vocab_size, emb_size, bias=False)
+    lin_layer.weight.data = emb.weight.data
+    return lin_layer
+
+@add_start_docstrings(
+    "The ProphetNet Model with a language modeling head. Can be used for summarization.", PROPHETNET_START_DOCSTRING
+)
+class ProphetNetForConditionalGeneration(ProphetNetPreTrainedModel):
+    base_model_prefix = "model"
+
+    def __init__(self, config: ProphetNetConfig):
+        super().__init__(config)
+        base_model = ProphetNetModel(config)
+        self.model = base_model
+        #self.padding_idx = config.pad_token_id
+        self.padding_idx = -100
+        self.disable_ngram_loss = config.disable_ngram_loss
+
+    def output_layer(self, features, **kwargs):
+        return F.linear(features, self.model.embed_tokens.weight)
+
+    @add_start_docstrings_to_callable(PROPHETNET_INPUTS_DOCSTRING)
+    def forward(
+        self,
+        input_ids,
+        attention_mask=None,
+        encoder_outputs=None,
+        decoder_input_ids=None,
+        decoder_cached_states=None,
+        labels=None,
+        use_cache=False,
+        output_attentions=None,
+        **unused,
+    ):
+        if "lm_labels" in unused:
+            warnings.warn(
+                "The `lm_labels` argument is deprecated and will be removed in a future version, use `labels` instead.",
+                DeprecationWarning,
+            )
+            labels = unused.pop("lm_labels")
+
+        outputs = self.model(
+            input_ids,
+            attention_mask=attention_mask,
+            decoder_input_ids=decoder_input_ids,
+            encoder_outputs=encoder_outputs,
+            decoder_cached_states=decoder_cached_states,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+        )
+        predicting_streams = outputs[0][1:]
+        predicting_streams_logits = [self.output_layer(x) for x in predicting_streams]
+        #lm_logits = F.linear(outputs[0], self.model.shared.weight, bias=self.final_logits_bias)
+
+        if labels is not None:
+            # fine-tune
+            expend_targets = labels.new_zeros(self.config.ngram, labels.size(0), labels.size(1)).fill_(self.padding_idx)
+            for i in range(self.config.ngram):
```

Comment:
can we try not to use the for look here? It should be quite easy to get rid of it here I think